core/src/core/pmmr.rs

```diff
@@ -322,48 +200,41 @@ where
 		res.expect("no root, invalid tree")
 	}
 
-	/// Build a Merkle proof for the element at the given position in the MMR
+	/// Build a Merkle proof for the element at the given position.
 	pub fn merkle_proof(&self, pos: u64) -> Result<MerkleProof, String> {
-		debug!(
-			LOGGER,
-			"merkle_proof (via rewind) - {}, last_pos {}", pos, self.last_pos
-		);
+		debug!(LOGGER, "merkle_proof  {}, last_pos {}", pos, self.last_pos);
 
 		if !is_leaf(pos) {
 			return Err(format!("not a leaf at pos {}", pos));
 		}
 
-		let root = self.root();
-
 		let node = self.get_hash(pos)
 			.ok_or(format!("no element at pos {}", pos))?;
 
 		let mmr_size = self.unpruned_size();
 
+		if mmr_size == 1 {
+			return Ok(MerkleProof {
+				mmr_size,
+				path: vec![node],
+			});
+		}
+
 		let family_branch = family_branch(pos, self.last_pos);
 
-		let path = family_branch
+		let mut path = family_branch
 			.iter()
-			.map(|x| (self.get_from_file(x.1).unwrap_or(Hash::default()), x.1))
+			.filter_map(|x| self.get_from_file(x.1))
```

Comment:
why filter_map instead of map ?

Reply:
So we don't need to do the extra `unwrap_or()` call to handle `None` values.
If there is nothing in the file at this pos then we filter it out and don't need to map it to anything.
This should never happen in practice, but the code is cleaner this way.
Either way we end up building an invalid Merkle proof if the file is actually missing hashes at the necessary pos.